examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart

```diff
@@ -0,0 +1,658 @@
+    // Update _internalList based on the model
+    _internalList = ScopedModel.of<AppStateModel>(context).productsInCart.keys.toList();
+    final Set<int> internalSet = Set<int>.from(_internalList);
+    final Set<int> listSet = Set<int>.from(_list.list);
+
+    final Set<int> difference = internalSet.difference(listSet);
+    if (difference.isEmpty) {
+      return;
+    }
+
+    for (int product in difference) {
+      if (_internalList.length < _list.length) {
+        _list.remove(product);
+      } else if (_internalList.length > _list.length) {
+        _list.add(product);
+      }
+    }
+
+    while (_internalList.length != _list.length) {
+      int index = 0;
+      // Check bounds and that the list elements are the same
+      while (_internalList.isNotEmpty &&
+          _list.length > 0 &&
+          index < _internalList.length &&
+          index < _list.length &&
+          _internalList[index] == _list[index]) {
+        index++;
+      }
+    }
+  }
+
+  Widget _buildAnimatedList() {
+    return AnimatedList(
+      key: _listKey,
+      shrinkWrap: true,
+      itemBuilder: _buildThumbnail,
+      initialItemCount: _list.length,
+      scrollDirection: Axis.horizontal,
+      physics: const NeverScrollableScrollPhysics(), // Cart shouldn't scroll
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    _updateLists();
+    return ScopedModelDescendant<AppStateModel>(
+      builder: (BuildContext context, Widget child, AppStateModel model) => _buildAnimatedList(),
+    );
+  }
+}
+
+class ExtraProductsNumber extends StatelessWidget {
+  // Calculates the number to be displayed at the end of the row if there are
+  // more than three products in the cart. This calculates overflow products,
+  // including their duplicates (but not duplicates of products shown as
+  // thumbnails).
+  int _calculateOverflow(AppStateModel model) {
+    final Map<int, int> productMap = model.productsInCart;
+    // List created to be able to access products by index instead of ID.
+    // Order is guaranteed because productsInCart returns a LinkedHashMap.
+    final List<int> products = productMap.keys.toList();
+    int overflow = 0;
+    final int numProducts = products.length;
+    if (numProducts > 3) {
+      for (int i = 3; i < numProducts; i++) {
+        overflow += productMap[products[i]];
+      }
+    }
+    return overflow;
+  }
+
+  Widget _buildOverflow(AppStateModel model, BuildContext context) {
+    if (model.productsInCart.length > 3) {
+      final int numOverflowProducts = _calculateOverflow(model);
+      // Maximum of 99 so padding doesn't get messy.
+      final int displayedOverflowProducts = numOverflowProducts <= 99 ? numOverflowProducts : 99;
+      return Container(
+        child: Text(
+          '+$displayedOverflowProducts',
+          style: Theme.of(context).primaryTextTheme.button,
+        ),
+      );
+    } else {
+      return Container(); // build() can never return null.
```

Comment:
return const SizedBox()

This method might be easier to read if the first line was:
```
if (model.productsInCart.length <= 3)
  return const SizedBox();
// no else clause
...
```